sgv2-service-common/src/main/java/io/stargate/sgv2/common/grpc/DefaultStargateBridgeClient.java

```diff
@@ -111,22 +100,66 @@
   }
 
   @Override
-  public CompletionStage<Optional<CqlKeyspaceDescribe>> getKeyspaceAsync(String keyspaceName) {
-    return authorizeSchemaReadAsync(SchemaReads.keyspace(keyspaceName, sourceApi))
-        .thenCompose(
-            authorized -> {
-              if (!authorized) {
-                throw new UnauthorizedKeyspaceException(keyspaceName);
+  public CompletionStage<Optional<CqlKeyspaceDescribe>> getKeyspaceAsync(
+      String keyspaceName, boolean checkIfAuthorized) {
+    if (checkIfAuthorized) {
+      return authorizeSchemaReadAsync(SchemaReads.keyspace(keyspaceName, sourceApi))
+          .thenCompose(
+              authorized -> {
+                if (!authorized) {
+                  throw new UnauthorizedKeyspaceException(keyspaceName);
+                }
+                return getAuthorizedKeyspace(keyspaceName).thenApply(Optional::ofNullable);
+              });
+    } else {
+      return getAuthorizedKeyspace(keyspaceName).thenApply(Optional::ofNullable);
+    }
+  }
+
+  private CompletionStage<CqlKeyspaceDescribe> getAuthorizedKeyspace(String keyspaceName) {
+    CompletableFuture<CqlKeyspaceDescribe> result = new CompletableFuture<>();
+
+    CqlKeyspaceDescribe cached = keyspaceCache.getIfPresent(keyspaceName);
+    Optional<Integer> cachedHash = Optional.ofNullable(cached).map(CqlKeyspaceDescribe::getHash);
+
+    // The result is cached locally, but we always hit the bridge anyway, to check if it has a more
+    // recent version.
+    fetchKeyspaceFromBridge(keyspaceName, cachedHash)
+        .whenComplete(
+            (fetched, error) -> {
+              if (error != null) {
+                result.completeExceptionally(error);
+              } else if (fetched == null) {
+                // Keyspace does not exist. Update the cache if we previously thought it did.
+                if (cached != null) {
+                  keyspaceCache.invalidate(keyspaceName);
+                }
```

Comment:
hmm are you sure you are not getting back the error on keyspace not existing, in my tests I saw that these calls do not end in `OK` status? Maybe you should move this to the `error` case?

Reply:
Yeah definitely better to handle it here, I changed it.